deepmd/pt/model/descriptor/dpa3.py

```diff
@@ -0,0 +1,572 @@
+        self.activation_function = activation_function
+
+        self.repflows = DescrptBlockRepflows(
+            self.repflow_args.e_rcut,
+            self.repflow_args.e_rcut_smth,
+            self.repflow_args.e_sel,
+            self.repflow_args.a_rcut,
+            self.repflow_args.a_rcut_smth,
+            self.repflow_args.a_sel,
+            ntypes,
+            nlayers=self.repflow_args.nlayers,
+            n_dim=self.repflow_args.n_dim,
+            e_dim=self.repflow_args.e_dim,
+            a_dim=self.repflow_args.a_dim,
+            a_compress_rate=self.repflow_args.a_compress_rate,
+            a_compress_e_rate=self.repflow_args.a_compress_e_rate,
+            a_compress_use_split=self.repflow_args.a_compress_use_split,
+            n_multi_edge_message=self.repflow_args.n_multi_edge_message,
+            axis_neuron=self.repflow_args.axis_neuron,
+            update_angle=self.repflow_args.update_angle,
+            activation_function=self.activation_function,
+            update_style=self.repflow_args.update_style,
+            update_residual=self.repflow_args.update_residual,
+            update_residual_init=self.repflow_args.update_residual_init,
+            fix_stat_std=self.repflow_args.fix_stat_std,
+            optim_update=self.repflow_args.optim_update,
+            exclude_types=exclude_types,
+            env_protection=env_protection,
+            precision=precision,
+            seed=child_seed(seed, 1),
+        )
+
+        self.use_econf_tebd = use_econf_tebd
+        self.use_tebd_bias = use_tebd_bias
+        self.type_map = type_map
+        self.tebd_dim = self.repflow_args.n_dim
+        self.type_embedding = TypeEmbedNet(
+            ntypes,
+            self.tebd_dim,
+            precision=precision,
+            seed=child_seed(seed, 2),
+            use_econf_tebd=self.use_econf_tebd,
+            use_tebd_bias=use_tebd_bias,
+            type_map=type_map,
+        )
+        self.concat_output_tebd = concat_output_tebd
+        self.precision = precision
+        self.prec = PRECISION_DICT[self.precision]
+        self.exclude_types = exclude_types
+        self.env_protection = env_protection
+        self.trainable = trainable
+
+        assert self.repflows.e_rcut >= self.repflows.a_rcut
+        assert self.repflows.e_sel >= self.repflows.a_sel
+
+        self.rcut = self.repflows.get_rcut()
+        self.rcut_smth = self.repflows.get_rcut_smth()
+        self.sel = self.repflows.get_sel()
+        self.ntypes = ntypes
+
+        # set trainable
+        for param in self.parameters():
+            param.requires_grad = trainable
+        self.compress = False
+
+    def get_rcut(self) -> float:
+        """Returns the cut-off radius."""
+        return self.rcut
+
+    def get_rcut_smth(self) -> float:
+        """Returns the radius where the neighbor information starts to smoothly decay to 0."""
+        return self.rcut_smth
```

Comment:
UTs are not covering some of the methods of the dpa3 class, as the codecov reported.

Reply:
I added some uts (such as test share_params, which is indeed missed by dpa3 only), but other uncovered lines are common in other descriptors. We may need separate PRs to add more universal tests.